### PR TITLE
Fix path of pickle files (broken so far)

### PR DESCRIPTION
**Pull request by jilljenn**
Dans la foulée j'ai vraiment cru que je pourrais faire un test de la route recommandation mais c'est galère parce qu'il faut :

- ajouter des œuvres bidon pour les recommandations, avec leurs ratings
- gérer le cas où un objet Pickle est déjà enregistré ou pas

En plus du coup ça va ralentir les tests. Notez qu'on peut juste tester MangakiZero, mais faut quand même les œuvres bidon.

### CODE REVIEW

File: mangaki/mangaki/utils/algo.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def fit_algo(algo_name, queryset, backup_filename):
     anonymized = dataset.make_anonymous_data(queryset)
     algo.set_parameters(anonymized.nb_users, anonymized.nb_works)
     algo.fit(anonymized.X, anonymized.y)
-    if algo_name in {'svd', 'als', 'wals'}:  # KNN is constantly refreshed
+    if algo_name in {'svd', 'als'}:  # KNN is constantly refreshed
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/algo.py]: Why WALS is removed ?

File: mangaki/mangaki/utils/common.py [hunk 2]
```diff
@@ -13,12 +13,20 @@ def __init__(self):
         self.nb_users = None
         self.nb_works = None
 
+    def get_backup_path(self, filename):
+        return os.path.join(PICKLE_DIR, self.get_backup_filename())
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/common.py]: filename is an unused parameter. Consider using @property or use the parameter.

File: mangaki/mangaki/utils/common.py [hunk 3]
```diff
@@ -13,12 +13,20 @@ def __init__(self):
...
+
+    def has_backup(self, filename=None):
+        if filename is None:
+            filename = self.get_backup_filename()
+        return os.path.isfile(self.get_backup_path(filename))
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/common.py]: What about : `filename or self.get_backup_filename()` ?

File: mangaki/mangaki/utils/common.py [hunk 4]
```diff
@@ -29,5 +37,8 @@ def set_parameters(self, nb_users, nb_works):
     def get_shortname(self):
         return 'algo'
 
+    def get_backup_filename(self):
+        return '%s.pickle' % self.get_shortname()
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/common.py]: Consider using @property.

File: mangaki/mangaki/utils/data.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import random
 from collections import Counter, namedtuple
 from mangaki.utils.values import rating_values
+from mangaki.utils.common import PICKLE_DIR
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/data.py]: IMHO, that should go in Django Settings.

File: mangaki/mangaki/utils/recommendations.py [hunk 6]
```diff
@@ -48,13 +48,12 @@ def get_reco_algo(user, algo_name='knn', category='all'):
...
 
     dataset = Dataset()
-    backup_filename = '%s.pickle' % algo_name
-    if os.path.isfile(os.path.join('pickles', backup_filename)):  # When Algo class will be there: 'if algo.has_backup():'
-        algo = ALGOS[algo_name]()
-        algo.load(backup_filename)
-        dataset.load('ratings-' + backup_filename)
+    algo = ALGOS[algo_name]()
+    if algo.has_backup():
+        algo.load(algo.get_backup_filename())
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/recommendations.py]: If it's frequent to use the default filename, consider use it as a default argument for the load method.

File: mangaki/mangaki/utils/algo.py (hunk 1 above)
Comment (jilljenn) [inline mangaki/mangaki/utils/algo.py]: The answer is in the commit title: “Since TensorFlow 1.0, MangakiWALS models cannot be saved anymore” ^^'

Comment (RaitoBezarius) [inline mangaki/mangaki/utils/algo.py]: That's definitely right. My bad for not reading your commit title! But you should have summarized it in your PR message, maybe?

Comment (jilljenn) [inline mangaki/mangaki/utils/algo.py]: It is now an issue :) https://github.com/mangaki/mangaki/issues/274

File: mangaki/mangaki/utils/common.py (hunk 2 above)
Comment (jilljenn) [inline mangaki/mangaki/utils/common.py]: Tu m'as fait peur. En fait ouais il faut faire attention de pas écraser.

File: mangaki/mangaki/utils/common.py (hunk 3 above)
Comment (jilljenn) [inline mangaki/mangaki/utils/common.py]: Trop long je pense.

File: mangaki/mangaki/utils/recommendations.py (hunk 6 above)
Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: Je pense juste que je ne peux pas mettre `algo.load()` parce que ça fait bizarre à lire, tu ne trouves pas ?

File: mangaki/mangaki/utils/data.py (hunk 5 above)
Comment (jilljenn) [inline mangaki/mangaki/utils/data.py]: OK mais please, futur commit lorsqu'on aura mis les tests.

File: mangaki/mangaki/utils/common.py (hunk 4 above)
Comment (jilljenn) [inline mangaki/mangaki/utils/common.py]: If it's a property, I should call it backup_filename, right? (without the get)

File: mangaki/mangaki/utils/recommendations.py (hunk 6 above)
Comment (Elarnon) [inline mangaki/mangaki/utils/recommendations.py]: Je pense que juste `algo.load()` est un peu clunky mais pourquoi pas; si il faut avoir une fonction qui fasse ça `algo.try_load_backup()` ou `algo.load_backup_if_exists()` est peut-être approprié. Je pense aussi que garder le `if has_backup` est OK for now.

File: mangaki/mangaki/utils/common.py (hunk 4 above)
Comment (Elarnon) [inline mangaki/mangaki/utils/common.py]: Yes. I think that either a function or a property is fine here.

**Comment by Elarnon** [comment]
I think this PR is a perfect example of what we talked about at the last meeting: the requested changes are mostly cosmetic and somewhat subjective, and thus I don't think they warrant a second round of review for the sake of moving forward.

@jilljenn @RaitoBezarius Would you agree that in cases such as this one (and especially when the changes come from an active community member - as opposed to a new contributor), it would be more productive to potentially make such comments when they provide value overall, but still accept the changes and let the author address some or all of them at their discretion? I think it would help prevent some PRs from becoming long pointless discussions about style, but happy to hear your thoughts on this.

**Comment by jilljenn** [comment]
Yes, that's great. Possibly the green check will motivate the PR's author to do as much extra work as they can.

(Or they will just click the big green button and completely ignore the useful messages.)

**Comment by RaitoBezarius** [comment]
@Elarnon I definitely agree with what you said, it's just a difficult habit of mine to comment on everything I can…
But, at least, I wish we put FIXME / TODO comments regarding some items so that we do not forget about what we should clean up after the PR is merged.

**Comment by Elarnon** [comment]
Maybe I explained myself poorly - the idea is *not* to prevent people from making comments they think improve the PR as a whole, but rather to try and make the PR process less resource-intensive in terms of developer time (which we have really little of).

If we look at the comments in this PR we have:
 - A question on why a small change was made
 - Three code style suggestions
 - A request to move a constant to the `settings` module

In my opinion, none of these are "blockers" for this code: they are valid comments, they are small and easy changes, and I think it would make sense to say "this is good for me as long as you make those changes (except if you find them unwarranted), I trust you from there", i.e. not request another round of review. If we find out that people end up being lazy and not addressing those comments, we can address the issue at that point. Btw, I will reiterate that I am proposing those process changes only for regular contributors - but in any case this GitHub issue is probably not the right place to discuss this, let's move the discussion to Slack.

**Comment by RaitoBezarius** [comment]
Alright, I didn't understand it this way, let's move the discussion on Slack.

**Pull request merged**